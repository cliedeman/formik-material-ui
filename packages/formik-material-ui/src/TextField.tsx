import * as React from 'react';
import MuiTextField, { TextFieldProps as MuiTextFieldProps } from '@material-ui/core/TextField';
import { FieldProps, getIn } from 'formik';

export interface TextFieldProps
  extends FieldProps,
    Omit<MuiTextFieldProps, 'name' | 'value' | 'error'> {
  touchOnChange?: boolean
}

export function fieldToTextField({
  disabled,
  field,
  form: { isSubmitting, touched, errors, setTouched },
  touchOnChange,
  ...props
}: TextFieldProps): MuiTextFieldProps {
  const fieldError = getIn(errors, field.name);
  const showError = getIn(touched, field.name) && !!fieldError;

  return {
    ...props,
    ...field,
    onChange(e) {
      if (touchOnChange && !getIn(touched, field.name)) {
        setTouched({ [field.name]: true });
      }
      const onChange = props.onChange || field.onChange;
      onChange && onChange(e);
    },
    error: showError,
    helperText: showError ? fieldError : props.helperText,
    disabled: disabled ?? isSubmitting,
    variant: props.variant,
  };
}

export function TextField({ children, ...props }: TextFieldProps) {
  return <MuiTextField {...fieldToTextField(props)}>{children}</MuiTextField>;
}

TextField.displayName = 'FormikMaterialUITextField';
