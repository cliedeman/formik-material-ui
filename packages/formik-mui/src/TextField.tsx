import * as React from 'react';
import {
  type TextFieldProps as MuiTextFieldProps,
  TextField as MuiTextField
} from '@mui/material';
import { type FieldProps, getIn } from 'formik';

export interface TextFieldProps
  extends FieldProps,
    Omit<MuiTextFieldProps, 'name' | 'value' | 'error'> {}

export function fieldToTextField({
  disabled,
  field: { onBlur: fieldOnBlur, ...field },
  form: { isSubmitting, touched, errors },
  onBlur,
  helperText,
  ...props
}: TextFieldProps): MuiTextFieldProps {
  const fieldError = getIn(errors, field.name);
  const showError = getIn(touched, field.name) && !!fieldError;

  return {
    error: showError,
    helperText: showError ? fieldError : helperText,
    disabled: disabled ?? isSubmitting,
    onBlur:
      onBlur ??
      function (e) {
        fieldOnBlur(e ?? field.name);
      },
    ...field,
    ...props,
  };
}

export function TextField({ children, ...props }: TextFieldProps) {
  return <MuiTextField {...fieldToTextField(props)}>{children}</MuiTextField>;
}

TextField.displayName = 'FormikMaterialUITextField';
