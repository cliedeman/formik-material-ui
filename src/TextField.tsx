// @ts-ignore
import * as React from 'react';
import MuiTextField, {
  TextFieldProps as MuiTextFieldProps,
} from '@material-ui/core/TextField';
import { FieldProps, getIn } from 'formik';

export interface TextFieldProps extends FieldProps, MuiTextFieldProps {}

const TextField: React.ComponentType<TextFieldProps> = ({
  field,
  form,
  disabled = false,
  ...props
}) => {
  const { name } = field;
  const { touched, errors, isSubmitting } = form;
  return (
    <MuiTextField
      {...props}
      {...field}
      error={getIn(touched, name) && !!getIn(errors, name)}
      helperText={
        getIn(touched, name) && getIn(errors, name)
          ? getIn(errors, name)
          : props.helperText
      }
      disabled={isSubmitting || disabled}
    />
  );
};

TextField.displayName = 'FormikMaterialUITextField';

export default TextField;
