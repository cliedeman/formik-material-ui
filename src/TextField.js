/* @flow */
import TextField from '@material-ui/core/TextField';

import createComponent from './createComponent';

export default createComponent(
  TextField,
  ({ field, form: { touched, errors, isSubmitting }, disabled = false, ...props }) => {
    const { name } = field;

    return {
      ...props,
      ...field,
      error: touched[name] && !!errors[name],
      helperText: errors[name] ? errors[name] : props.helperText,
      disabled: isSubmitting || disabled,
    };
  }
);
