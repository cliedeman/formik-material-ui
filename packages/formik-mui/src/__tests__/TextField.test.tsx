import * as React from 'react';
import { test, expect } from 'vitest';
import { Formik, Form, Field } from 'formik';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { render } from '@testing-library/react'

import { TextField } from '../TextField';

test('TextField Renders Correctly', () => {
  const { asFragment } = render(
    <Formik initialValues={{}} onSubmit={() => {}}>
      <Form>
        <Field component={TextField} name="test" label="Text" />
      </Form>
    </Formik>
  );

  expect(asFragment()).toMatchSnapshot();
});

test('Outlined TextField', () => {
  const { asFragment } = render(
    <Formik initialValues={{}} onSubmit={() => {}}>
      <Form>
        <Field
          component={TextField}
          name="test"
          label="Text"
          variant="outlined"
        />
      </Form>
    </Formik>
  );

  expect(asFragment()).toMatchSnapshot();
});

test('Override TextField Theme Variant', () => {
  const theme = createTheme({});

  const { asFragment } = render(
    <ThemeProvider theme={theme}>
      <Formik initialValues={{}} onSubmit={() => {}}>
        <Form>
          <Field component={TextField} name="test" label="Text" />
        </Form>
      </Formik>
    </ThemeProvider>
  );

  expect(asFragment()).toMatchSnapshot();
});
