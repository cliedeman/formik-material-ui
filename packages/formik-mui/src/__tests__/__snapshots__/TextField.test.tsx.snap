// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Outlined TextField 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r1:"
        id=":r1:-label"
      >
        Text
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r1:"
          name="test"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Text
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Override TextField Theme Variant 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r2:"
        id=":r2:-label"
      >
        Text
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r2:"
          name="test"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Text
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`TextField Renders Correctly 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r0:"
        id=":r0:-label"
      >
        Text
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          name="test"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Text
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
